cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -0,0 +1,225 @@
+
+namespace {    // anonymous
+
+template <typename T, typename Filter, bool has_validity>
+__global__
+void copy_if_else_kernel(  column_device_view const lhs,
+                           column_device_view const rhs,
+                           Filter filter,
+                           mutable_column_device_view out,
+                           cudf::size_type * __restrict__ const null_count)
+{   
+   const cudf::size_type tid = threadIdx.x + blockIdx.x * blockDim.x;   
+   const int w_id = tid / warp_size;
+   // begin/end indices for the column data
+   cudf::size_type begin = 0;
+   cudf::size_type end = lhs.size();   
+   // warp indices.  since 1 warp == 32 threads == sizeof(bit_mask_t) * 8,
+   // each warp will process one (32 bit) of the validity mask via 
+   // __ballot_sync()
+   cudf::size_type w_begin = cudf::util::detail::bit_container_index<bit_mask::bit_mask_t>(begin);
+   cudf::size_type w_end = cudf::util::detail::bit_container_index<bit_mask::bit_mask_t>(end);
+
+   // lane id within the current warp
+   const int w_lane_id = threadIdx.x % warp_size;
+
+   // store a null count for each warp in the block
+   constexpr cudf::size_type b_max_warps = 32;
+   __shared__ uint32_t b_warp_null_count[b_max_warps];
+   // initialize count to 0. we have to do this because the WarpReduce
+   // at the end will end up summing all values, even ones which we never 
+   // visit.
+   if(has_validity){   
+      if(threadIdx.x < b_max_warps){
+         b_warp_null_count[threadIdx.x] = 0;
+      }   
+      __syncthreads();   
+   }   
+
+   // current warp.
+   cudf::size_type w_cur = w_begin + w_id;         
+   // process each grid
+   while(w_cur <= w_end){
+      // absolute element index
+      cudf::size_type index = (w_cur * warp_size) + w_lane_id;
+      bool in_range = (index >= begin && index < end);
+
+      bool valid = true;
+      if(has_validity){
+         valid = in_range && filter(index) ? lhs.is_valid(index) : rhs.is_valid(index);
+      }
+
+      // do the copy if-else, but only if this element is valid in the column to be copied 
+      if(in_range && valid){ 
+         out.element<T>(index) = filter(index) ? lhs.element<T>(index) : rhs.element<T>(index);
+      }
+      
+      // update validity
+      if(has_validity){
+         // get mask indicating which threads in the warp are actually in range
+         int w_active_mask = __ballot_sync(0xFFFFFFFF, in_range);      
+         // the final validity mask for this warp
+         int w_mask = __ballot_sync(w_active_mask, valid);
+         // only one guy in the warp needs to update the mask and count
+         if(w_lane_id == 0){
+            out.set_mask_word(w_cur, w_mask);
+            cudf::size_type b_warp_cur = threadIdx.x / warp_size;
+            b_warp_null_count[b_warp_cur] = __popc(~(w_mask | ~w_active_mask));
+         }
+      }      
+
+      // next grid
+      w_cur += blockDim.x * gridDim.x;
+   }
+
+   if(has_validity){
+      __syncthreads();
```

Comment:
These lines duplicate `cudf::experimental::detail::single_lane_block_sum_reduce()` 

```suggestion
      size_type b_null_count = single_lane_block_sum_reduce(w_null_count);
      
      // only one thread in the warp needs to do the actual store
      if(w_lane_id == 0){
          // using an atomic here because there are multiple blocks doing this work
          atomicAdd(null_count, b_null_count);
      }
```

Note that you can get rid of the shared memory completely. Just store `w_null_count` in a register variable.  `single_lane_...` does the rest for you.

Reply:
Try to mimic the exact same pattern used here: https://github.com/rapidsai/cudf/blob/1a16dec10840157f74a75fe48e0b0605d82d1b0f/cpp/include/cudf/detail/valid_if.cuh#L58